plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/reader/TrinoReaderContext.java

```diff
@@ -0,0 +1,159 @@
+        super(tableConfig, new AvroJavaTypeConverter());
+    }
+
+    @Override
+    public HoodieRecord<IndexedRecord> constructHoodieRecord(BufferedRecord<IndexedRecord> bufferedRecord, String partitionPath)
+    {
+        if (bufferedRecord.isDelete()) {
+            return new HoodieEmptyRecord<>(
+                    new HoodieKey(bufferedRecord.getRecordKey(), partitionPath),
+                    HoodieRecord.HoodieRecordType.AVRO);
+        }
+
+        return new HoodieAvroIndexedRecord(bufferedRecord.getRecord());
+    }
+
+    public static Object getFieldValueFromIndexedRecord(
+            IndexedRecord record,
+            String fieldName) {
+        Schema currentSchema = record.getSchema();
+        IndexedRecord currentRecord = record;
+        String[] path = fieldName.split("\\.");
+        for (int i = 0; i < path.length; i++) {
+            if (currentSchema.isUnion()) {
+                currentSchema = AvroSchemaUtils.resolveNullableSchema(currentSchema);
+            }
+            Schema.Field field = currentSchema.getField(path[i]);
+            if (field == null) {
+                return null;
+            }
+            Object value = currentRecord.get(field.pos());
+            if (i == path.length - 1) {
+                return value;
+            }
+            currentSchema = field.schema();
+            currentRecord = (IndexedRecord) value;
+        }
+        return null;
+    }
+
+    @Override
+    public IndexedRecord mergeWithEngineRecord(Schema schema, Map<Integer, Object> updateValues, BufferedRecord<IndexedRecord> baseRecord)
+    {
+        IndexedRecord engineRecord = baseRecord.getRecord();
+        for (Map.Entry<Integer, Object> value : updateValues.entrySet()) {
+            engineRecord.put(value.getKey(), value.getValue());
+        }
+        return engineRecord;
+    }
+
+    @Override
+    public IndexedRecord constructEngineRecord(Schema recordSchema, Object[] fieldValues)
+    {
+        GenericData.Record record = new GenericData.Record(recordSchema);
+        for (int i = 0; i < fieldValues.length; i++) {
+            record.put(i, fieldValues[i]);
+        }
+        return record;
+    }
+
+    @Override
+    public Object getValue(IndexedRecord record, Schema schema, String fieldName)
+    {
+        return getFieldValueFromIndexedRecord(record, fieldName);
+    }
+
+    @Override
+    public String getMetaFieldValue(IndexedRecord record, int pos)
+    {
+        return record.get(pos).toString();
+    }
+
+    @Override
+    public IndexedRecord convertAvroRecord(IndexedRecord avroRecord)
+    {
+        return avroRecord;
+    }
+
+    @Override
+    public GenericRecord convertToAvroRecord(IndexedRecord record, Schema schema)
+    {
+        GenericRecord genericRecord = new GenericData.Record(schema);
+        for (Schema.Field field : schema.getFields()) {
+            genericRecord.put(field.name(), record.get(field.pos()));
+        }
+        return genericRecord;
```

Comment:
`IndexedRecord` can be casted to `GenericRecord`, so no need to reconstruct the record which introduces overhead?